File: app/src/main/java/ua/syt0r/furiganit/viewmodel/service/ServiceManagerViewModel.java
package ua.syt0r.furiganit.viewmodel.service;

import android.app.Application;
import android.provider.Settings;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.OnLifecycleEvent;

import ua.syt0r.furiganit.model.repository.status.ServiceStatus;
import ua.syt0r.furiganit.model.repository.status.ServiceStatusRepository;

public class ServiceManagerViewModel extends AndroidViewModel {

    public enum ServiceManagerStates { CANT_DRAW_OVERLAY, STOPPED, LAUNCHING, RUNNING }

    private ServiceStatusRepository serviceStatusRepository;

    private MediatorLiveData<ServiceManagerStates> mutableState = new MediatorLiveData<>();

    public ServiceManagerViewModel(@NonNull Application application) {
        super(application);
        serviceStatusRepository = new ServiceStatusRepository(application);
        mutableState.addSource(serviceStatusRepository.subscribeOnStatus(), status -> {
            updateState();
        });
        updateState();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        serviceStatusRepository.unsubscribe();
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    void onResume() { updateState(); }

    private boolean canDrawOverlay() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M)
            return Settings.canDrawOverlays(getApplication());
        return true;
    }

    private void updateState() {

        ServiceStatus status = serviceStatusRepository.subscribeOnStatus().getValue();
        if (status == null) status = ServiceStatus.STOPPED;

        if (canDrawOverlay())
            switch (status) {
                case STOPPED:
                    mutableState.setValue(ServiceManagerStates.STOPPED);
                    break;
                case LAUNCHING:
                    mutableState.setValue(ServiceManagerStates.LAUNCHING);
                    break;
                case RUNNING:
                    mutableState.setValue(ServiceManagerStates.RUNNING);
            }
        else
            mutableState.setValue(ServiceManagerStates.CANT_DRAW_OVERLAY);

    }


    public LiveData<ServiceManagerStates> subscribeOnServiceStatus() {
        return mutableState;
    }

}

File: app/src/main/java/ua/syt0r/furiganit/app/main/screens/service_manager/ServiceManagerViewModel.kt
package ua.syt0r.furiganit.app.main.screens.service_manager

import androidx.lifecycle.*
import ua.syt0r.furiganit.core.overlay.OverlayManager
import ua.syt0r.furiganit.core.service.ServiceState
import ua.syt0r.furiganit.core.service.ServiceStateObservable

class ServiceManagerViewModel(
    private val serviceStatusObservable: ServiceStateObservable
) : ViewModel(), LifecycleObserver {

    private val mutableState = MediatorLiveData<ServiceManagerState>()

    init {
        mutableState.addSource(serviceStatusObservable.subscribeToStateChanges()) { updateState() }
        updateState()
    }

    override fun onCleared() {
        super.onCleared()
        serviceStatusObservable.unsubscribe()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    private fun updateState() {
        val serviceState = serviceStatusObservable.subscribeToStateChanges().value
            ?: ServiceState.STOPPED

        mutableState.value = serviceState.toServiceManagerState()
    }

    fun subscribeOnServiceState(): LiveData<ServiceManagerState> {
        return mutableState
    }

}
